src/frameworks/compose/discover/runtime/node.ts

```diff
@@ -0,0 +1,212 @@
+      this.runtimeRequiredFiles.map((file) => fs.exists(file))
+    );
+
+    return areAllFilesPresent.every((present) => present);
+  }
+
+  getRuntimeName(): string {
+    return NODE_RUNTIME_ID;
+  }
+
+  getNodeImage(engine: Record<string, string> | undefined): string {
+    // If no version is mentioned explicitly, assuming application is compatible with latest version.
+    if (!engine || !engine.node) {
+      return `node:${supportedNodeVersions[supportedNodeVersions.length - 1]}-slim`;
+    }
+    const versionNumber = engine.node;
+
+    if (!supportedNodeVersions.includes(versionNumber)) {
+      throw new FirebaseError(
+        `This integration expects Node version ${conjoinOptions(
+          supportedNodeVersions,
+          "or"
+        )}. You're running version ${versionNumber}, which is not compatible.`
+      );
+    }
+
+    return `node:${versionNumber}-slim`;
+  }
+
+  async getPackageManager(fs: FileSystem): Promise<PackageManager> {
+    try {
+      if (await fs.exists(YARN_LOCK)) {
+        return "yarn";
+      }
+
+      return "npm";
+    } catch (error: any) {
+      logger.error("Failed to check files to identify package manager");
+      throw error;
+    }
+  }
+
+  getDependencies(packageJSON: PackageJSON): Record<string, string> {
+    return { ...packageJSON.dependencies, ...packageJSON.devDependencies };
+  }
+
+  packageManagerInstallCommand(packageManager: PackageManager): string | undefined {
+    const packages: string[] = [];
+    if (packageManager === "yarn") {
+      packages.push("yarn");
+    }
+    if (!packages.length) {
+      return undefined;
+    }
+
+    return `npm install --global ${packages.join(" ")}`;
+  }
+
+  installCommand(fs: FileSystem, packageManager: PackageManager): string {
+    let installCmd = "npm install";
+
+    if (packageManager === "yarn") {
+      installCmd = "yarn install";
+    }
+
+    return installCmd;
+  }
+
+  async detectedCommands(
+    packageManager: PackageManager,
+    scripts: Record<string, string> | undefined,
+    matchedFramework: FrameworkSpec | null,
+    fs: FileSystem
+  ): Promise<LifecycleCommands> {
+    return {
+      build: this.getBuildCommand(packageManager, scripts, matchedFramework),
+      dev: this.getDevCommand(packageManager, scripts, matchedFramework),
+      run: await this.getRunCommand(packageManager, scripts, matchedFramework, fs),
+    };
+  }
+
+  executeScript(packageManager: string, scriptName: string): string {
+    return `${packageManager} run ${scriptName}`;
+  }
+
+  executeFrameworkCommand(packageManager: PackageManager, command: Command): Command {
+    if (packageManager === "npm" || packageManager === "yarn") {
+      command.cmd = "npx " + command.cmd;
+    }
+
+    return command;
+  }
+
+  getBuildCommand(
+    packageManager: PackageManager,
+    scripts: Record<string, string> | undefined,
+    matchedFramework: FrameworkSpec | null
+  ): Command {
+    let buildCommand: Command = { cmd: "" };
```

Comment:
I think I'd prefer returning `undefined` if the command is empty to denote that we have no build command. Do you think it's better to return an empty string instead?

Reply:
I think undefined is better than empty string as we can check if command is available or not.